src/Microsoft.TemplateSearch.TemplateDiscovery/PackChecking/PackSourceChecker.cs

```diff
@@ -30,65 +30,89 @@ public PackSourceChecker(IPackProvider packProvider, PackPreFilterer packPreFilt
         public async Task<PackSourceCheckResult> CheckPackagesAsync()
         {
             List<PackCheckResult> checkResultList = new List<PackCheckResult>();
+            HashSet<IPackInfo> alreadySeenPacks = new HashSet<IPackInfo>();
             HashSet<string> alreadySeenTemplateIdentities = new HashSet<string>();
 
-            int count = 0;
-            Console.WriteLine($"{await _packProvider.GetPackageCountAsync().ConfigureAwait(false)} packs discovered, starting processing");
-
-            await foreach (IInstalledPackInfo packInfo in _packProvider.GetCandidatePacksAsync().ConfigureAwait(false))
+            foreach (IPackProvider packProvider in _packProviders)
             {
-                PackCheckResult preFilterResult = PrefilterPackInfo(packInfo);
-                if (preFilterResult.PreFilterResults.ShouldBeFiltered)
-                {
-                    Verbose.WriteLine($"{packInfo.Id}::{packInfo.Version} is skipped, {preFilterResult.PreFilterResults.Reason}");
-                    checkResultList.Add(preFilterResult);
-                }
-                else
+                Console.WriteLine($"Processing pack provider {packProvider.Name}:");
+                int count = 0;
+                Console.WriteLine($"{await packProvider.GetPackageCountAsync().ConfigureAwait(false)} packs discovered, starting processing");
+
+                await foreach (IPackInfo sourceInfo in packProvider.GetCandidatePacksAsync().ConfigureAwait(false))
                 {
-                    PackCheckResult packCheckResult = _packChecker.TryGetTemplatesInPack(packInfo, _additionalDataProducers, alreadySeenTemplateIdentities);
-                    checkResultList.Add(packCheckResult);
+                    if (alreadySeenPacks.Contains(sourceInfo))
+                    {
+                        Verbose.WriteLine($"Package {sourceInfo.Id}::{sourceInfo.Version} is already processed.");
+                        continue;
+                    }
+                    alreadySeenPacks.Add(sourceInfo);
 
+                    IDownloadedPackInfo packInfo = await packProvider.DownloadPackageAsync(sourceInfo).ConfigureAwait(false);
```

Comment:
As far as I know, NuGet client allows downloading directly into memory without writing onto disk. If we are planning to delete the packages anyway, could we do this all in memory?